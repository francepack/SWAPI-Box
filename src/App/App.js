import React, { Component } from 'react'
import '../Styles/index.scss'
import { makeFetch } from '../Fetch/Fetch'
import Header from '../Header/Header'
import Quote from '../Quote/Quote'
import Control from '../Control/Control'
import CardContainer from '../CardContainer/CardContainer'

export default class App extends Component {
  constructor() {
    super()
    this.state= {
      movieQuote: {},
      people: [],
      planets: [],
      vehicles: [],
      favorites: [],
      display: '',
      error: ''
    }
  }

  componentDidMount() {
    this.getMovieQuote();
  }

  getMovieQuote = () => {
    const movieNumber = Math.ceil(Math.random() * Math.floor(7));
    const quoteUrl = 'https://swapi.co/api/films/' + movieNumber;
    makeFetch(quoteUrl)
      .then(result => this.setState({ movieQuote: {quote: result.opening_crawl, title: result.title, date: result.release_date} }))
      .catch(error => this.setState({ error: error.message }))
  }

  getPeople = () => {
    if (!this.state.people.length) {
      const peopleUrl = 'https://swapi.co/api/people'
      makeFetch(peopleUrl)
        .then(people => this.addHomeworld(people))
        .then(people => this.addSpecies(people))
        .then(people => this.finalizeData(people))
        .then(people => this.setState({ people: people, display: 'people' }))
        .catch(error => this.setState({ error: error.message }))
    }
  }

  addHomeworld(people) {
    const addedHomeworld = people.results.map(person => {
      return makeFetch(person.homeworld)
        .then(result => ({...person, homeworld: result.name, population: result.population}))
    })
    return Promise.all(addedHomeworld)
  }

  addSpecies(people) {
    const addedSpecies = people.map(person => {
      return makeFetch(person.species)
        .then(result => ({...person, species: result.name, category:'person'}))
    })
    return Promise.all(addedSpecies)  
  }

  //Should I clean my info? More neat to do that in my fetch builds, or outside function?
  //Wondering if I need to normalize what I pass down, or have the card show a line for each key it sees

    finalizeData(arr) {
    const finishedObjects = arr.map(obj => {
      switch (obj.category) {
        case 'person':
          return { name: obj.name, 
            category: obj.category, 
            homeworld: obj.homeworld, 
            population: obj.population, 
            favorite: false, 
            id: Date.now()
          }
        break;
        case 'planet':
          return { name: obj.name, 
            category: obj.category, 
            terrain: obj.terrain, 
            population: obj.population, 
            climate: obj.climate, 
            residents: obj.residents, 
            favorite: false, 
            id: Date.now()
          }
        break;
        case 'vehicle':
          return { name: obj.name, 
            category: obj.category, 
            model: obj.model, 
            class: obj.vehicle_class, 
            passengers: obj.passengers, 
            favorite: false, 
            id: Date.now()
          }
        break;
      }  
    })
    return finishedObjects;
  }

  getPlanets = () => {
    if (!this.state.planets.length) {
      const planetUrl = 'https://swapi.co/api/planets'
      makeFetch(planetUrl)
        .then(planets => this.addResidents(planets))
        .then(planets => this.finalizeData(planets))
        .then(planets => this.setState({ planets: planets, display: 'planets'}))
        .catch(error => this.setState({ error: error.message }))
    } else {
      this.setState({ display: 'planets' });
    }
  }

  addResidents(planets) {
    const addedResidents = planets.results.map(planet => {
      return this.getResidents(planet)
        .then(result => ({...planet, residents: result, category: 'planet' }))
    })
    return Promise.all(addedResidents)
  }

  getResidents(planet) {
    const residents = planet.residents.map(resident => {
      return makeFetch(resident)
        .then(result => (result.name))
    })
    return Promise.all(residents)
  }

  getVehicles = () => {
    if (!this.state.vehicles.length) {
      const vehicleUrl = 'https://swapi.co/api/vehicles'
      makeFetch(vehicleUrl)
        .then(vehicles => (vehicles.results.map(vehicle => ({...vehicle, category: 'vehicle' }))))
        .then(vehicles => this.finalizeData(vehicles))
        .then(vehicles => this.setState({ vehicles: vehicles, display: 'vehicles' }))
        .catch(error => this.setState({ error: error.message }))
    } else {
      this.setState({ display: 'vehicles' });
    }
  }

  viewFavorites = () => {
    this.setState({ display: 'favorites' })
  }

  toggleFavorite = () => {

  }

  render() {
    console.log(this.state.people)
    console.log(this.state.display)
    console.log(this.state.planets)
    console.log(this.state.vehicles)


    const { display, error } = this.state;
    let cardDisplay;

    if (error) {
      cardDisplay = <p>{error}</p>
    } else if (!display) {
      cardDisplay = <p>Select Cool stuff to View</p>
    } else {
      cardDisplay = <CardContainer data={this.state[display]}
                    />
    }

    return (
      <div className="App">
        <Header viewFavorites={this.viewFavorites}
                favCount={this.state.favorites.length}
        />
        <Quote message={this.state.movieQuote} />
        <Control getPeople={this.getPeople}
                 getPlanets={this.getPlanets}
                 getVehicles={this.getVehicles}           
        />
        {cardDisplay}
      </div>
    );
  }
}