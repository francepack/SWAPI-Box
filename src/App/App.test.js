import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
// import { makeFetch } from '../Fetch/Fetch'
import { shallow } from 'enzyme';


describe("App", () => {
  let wrapper;
  let mockUrl;
  let mockData;
  let makeFetch;
  beforeEach(() => {
    wrapper = shallow(
      <App />
    )
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have default states", () => {
    expect(wrapper.state()).toEqual({
      movieQuote: {}, 
      people: [], 
      planets: [],
      vehicles: [],
      favorites: [],
      display: '',
      error: ''
    });
  });

  describe("getMovieQuote", () => {
    // it("should call makeFetch", () => {
    //   wrapper.instance().getMovieQuote(mockUrl)
    //   expect(makeFetch).toBeCalled()
    // });
    it("should update movieQuote state", () => {
      const mockMovie = {opening_crawl: 'Galaxy', title: 'A No Hope', release_date: 'Decenever 42, 1901'}
      const resultMovieQuote = {quote: 'Galaxy', title: 'A No Hope', date: 'Decenever 42, 1901'}
      makeFetch = jest.fn(() => (mockMovie));
      expect(wrapper.state("movieQuote")).toEqual({});
      wrapper.instance().getMovieQuote();
      expect(wrapper.state("movieQuote")).toEqual(resultMovieQuote);
    });

  });
  describe("getPeople", () => {
    // it("should call makeFetch", () => {

    // });
    it("should call addSpecies", () => {
      
    });
    it("should call finalizeData", () => {
      
    });
    it("should update state people and display", () => {
      
    });
  });
  describe('addHomeworld', () => {
    it("should add properties of homeworld and population to each obj in array", async () => {
      mockData = {results: [{name: 'Mason'}, {name: 'Isaac'}]};
      let fetchData = {name: 'Mars', population: 100000};
      let endData = [{name: 'Mason', name: 'Mars', population: 100000}, {name: 'Isaac'}];
      makeFetch = jest.fn(() => fetchData)
      const people = await wrapper.instance().addHomeworld(mockData)
      expect(people).toBe(endData)
    });
  });
  describe('addSpecies', () => {

  });
  describe('getPlanets', () => {
    it("should call makeFetch", () => {
      
    });
    it("should call addResidents", () => {
      
    });
    it("should update state planets and display", () => {
      
    });
  });
  describe('addResidents', () => {

  });
  describe('getResidents', () => {

  });
  describe('getVehicles', () => {
    it("should call makeFetch", () => {
      
    });
    it("should call finalizeData", () => {
      
    });
    it("should update state vehicles and display", () => {
      
    })
  });
  describe('finalizeData', () => {

  });
  describe('', () => {

  });
});  
