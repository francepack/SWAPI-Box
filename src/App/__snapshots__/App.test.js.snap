// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header
          favCount={0}
          viewFavorites={[Function]}
        />,
        <Quote
          message={Object {}}
        />,
        <Control
          getPeople={[Function]}
          getPlanets={[Function]}
          getVehicles={[Function]}
        />,
        <p>
          Select Cool stuff to View
        </p>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "favCount": 0,
          "viewFavorites": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "getPeople": [Function],
          "getPlanets": [Function],
          "getVehicles": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Select Cool stuff to View",
        },
        "ref": null,
        "rendered": "Select Cool stuff to View",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header
            favCount={0}
            viewFavorites={[Function]}
          />,
          <Quote
            message={Object {}}
          />,
          <Control
            getPeople={[Function]}
            getPlanets={[Function]}
            getVehicles={[Function]}
          />,
          <p>
            Select Cool stuff to View
          </p>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "favCount": 0,
            "viewFavorites": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "getPeople": [Function],
            "getPlanets": [Function],
            "getVehicles": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Select Cool stuff to View",
          },
          "ref": null,
          "rendered": "Select Cool stuff to View",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
